### PR TITLE
fix #1313  Remove read only isolation level

### LINKED ISSUE #1313: Remove IsolationLevelType::READ_ONLY
The way that we are marking that a txn is read-only is by setting it's isolation level to `READ_ONLY`.

https://github.com/cmu-db/peloton/blob/master/src/include/common/internal_types.h#L425

We currently don't support setting a txn to be read-only from the SQL interface, but the `TransactionManager` is checking for this `IsolationLevel` all over the place:

https://github.com/cmu-db/peloton/blob/master/src/concurrency/transaction_manager.cpp#L37

This is incorrect logic. `READ_ONLY` is not a valid isolation level. A txn can be different isolation levels and also read-only.

We need to remove  `IsolationLevelType::READ_ONLY` and then add a `readonly_` flag to the `TransactionContext` class. The `TransactionManager` should check this flag not the isolation type.

### PR DESCRIPTION
**Pull request by Nov11**
this is to #1313 

Details:
1.remove IsolationLevelType::READ_ONLY
2.add a flag in transactioncontext 
3.re-implement 'IsReadOnly' function. This is safe because the function has not been call anywhere before this modification.
4.substitute every usage of 'READ_ONLY' with IsReadOnly function call
5.add a new parameter to BeginTransaction and provide a short cut 'BeginTransactionReadOnly'
6.one invocation of BeginTransaction(READ_ONLY) is changed to 'BeginTransactionReadOnly(IsolationLevelType::READ_COMMITTED);'

### CODE REVIEW

**Comment by poojanilangekar** [comment]
@Nov11 Your change is breaking the `serializable_transaction_test `. Please fix it and let me know so I can review the change.

**Comment by Nov11** [comment]
It won't break tests now.
jenkins failed because of network error : 'stderr: fatal: unable to access 'https://github.com/cmu-db/peloton.git/': Failed to connect to github.com port 443: No route to host
'

**Comment by crd477** [comment]
On 2018-04-21 13:08, Lin Ma wrote:
> Hi @crd477 <https://github.com/crd477> @pervazea 
> <https://github.com/pervazea>, it looks like some of our PRs including 
> this one cannot download the repo with the Jenkins build again. For example:
> http://jenkins.db.cs.cmu.edu:8080/blue/organizations/jenkins/peloton/detail/PR-1322/1/pipeline/
> http://jenkins.db.cs.cmu.edu:8080/blue/organizations/jenkins/peloton/detail/PR-1321/1/pipeline/
> http://jenkins.db.cs.cmu.edu:8080/blue/organizations/jenkins/peloton/detail/PR-1133/86/pipeline/
> http://jenkins.db.cs.cmu.edu:8080/blue/organizations/jenkins/peloton/detail/PR-1310/7/pipeline/
> 
> I'm wondering is this because we have another machine that has 
> networking problem? Would really appreciate your help.
> 

Ugh...

Yes, this was the default route problem again that originally cropped up 
in #1295.  I thought I'd fixed it the other day but obviously it 
happened again.  I've implemented a more permanent fix now and it should 
not happen again.

Sorry (again) for the inconvenience...

File: src/concurrency/transaction_manager.cpp [hunk 1]
```diff
@@ -66,6 +60,10 @@ TransactionContext *TransactionManager::BeginTransaction(
     txn = new TransactionContext(thread_id, type, read_id);
   }
 
+  if(read_only){
```
Comment (apavlo) [inline src/concurrency/transaction_manager.cpp]: Fix spacing

File: src/include/concurrency/transaction_context.h [hunk 2]
```diff
@@ -269,7 +269,15 @@ class TransactionContext : public Printable {
...
   inline bool IsReadOnly() const {
-    return is_written_ == false && insert_count_ == 0;
+    return read_only_;
+  }
+
+  /**
+   * @brief      mark this context as read only
+   *
+   */
+  inline void setReadOnly() {
```
Comment (apavlo) [inline src/include/concurrency/transaction_context.h]: Remove inline

File: src/include/concurrency/transaction_context.h [hunk 3]
```diff
@@ -269,7 +269,15 @@ class TransactionContext : public Printable {
    * @return     True if read only, False otherwise.
    */
   inline bool IsReadOnly() const {
```
Comment (apavlo) [inline src/include/concurrency/transaction_context.h]: Remove inline

File: src/include/concurrency/transaction_context.h [hunk 4]
```diff
@@ -341,6 +349,8 @@ class TransactionContext : public Printable {
   IsolationLevelType isolation_level_;
 
   std::unique_ptr<trigger::TriggerSet> on_commit_triggers_;
+
+  bool read_only_ = false;
```
Comment (apavlo) [inline src/include/concurrency/transaction_context.h]: Add comment

File: src/include/concurrency/transaction_manager.h [hunk 5]
```diff
@@ -219,11 +219,16 @@ class TransactionManager {
   }
 
   TransactionContext *BeginTransaction(const IsolationLevelType type) {
-    return BeginTransaction(0, type);
+    return BeginTransaction(0, type, false);
+  }
+
+  TransactionContext *BeginTransactionReadOnly(const IsolationLevelType type) {
```
Comment (apavlo) [inline src/include/concurrency/transaction_manager.h]: This is unnecessary bloat. I would remove this.

**Comment by Nov11** [comment]
As there's already a read only transaction test prior to this change(SerializableTransactionTests->ReadOnlyTransactionTest), I thought it's adequate and didn't add new ones.
I broke that test case once.  poojanilangekar  reminded me of it. And it's been fixed.
If that test is not suffice to peloton, I'd like to do the addition.

**Comment by apavlo** [comment]
@Nov11 What does the test do? We need to test the following scenario:

```
Txn #1 | Txn #2
-------------------
BEGIN  |
W(X)   |
       | BEGIN R/O
       | R(X)
W(X)   |
COMMIT | 
       | R(X)
       | COMMIT
```

**Comment by apavlo** [comment]
@Nov11 Your comment about the broken test case looks to be correct. We want to test the scenario when there are two active txns running simultaneously. The code in `ReadOnlyTransactionTest` looks to have only one txn at a time.

Don't worry about any isolation level other than snapshot isolation.

**Comment by Nov11** [comment]
There's a concurrent test case now.

**Comment by poojanilangekar** [comment]
@apavlo @pervazea Can you please review this PR? 
@mengranwo has added support for change column name but we currently unable to test it because we need concurrent transactions to check the correctness.

**Comment by apavlo** [comment]
I will look at this today.

**Comment by apavlo** [comment]
I am going to merge this when the build passes. There are some stylistic things that I want to change but I am just going to fix that myself.

@Nov11 Thanks for your help!

**Pull request merged**